### PR TITLE
KAFKA-8058: Fix ConnectClusterStateImpl.connectors() method

### PR DESCRIPTION
**Pull request by C0urante**
This makes the `ConnectClusterStateImpl.connectors()` method synchronous, whereas before it was implicitly asynchronous with no way to tell whether it had completed or not.

More detail can be found in the [Jira](https://issues.apache.org/jira/browse/KAFKA-8058).

~Tested manually. Seems like a light enough change that even unit tests would be overkill, but if reviewers feel differently tests can be added.~
A unit test has also been added for the `ConnectClusterStateImpl` class.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by C0urante** [comment]
@mageshn @wicknicks @rayokota could you guys take a look when you have a chance? 😃

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java
Comment (wicknicks) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java]: nit: should we log an error if this method returns false?

**Review by wicknicks — COMMENTED** [review]
good catch, @C0urante. one comment.

**Review by mageshn — APPROVED** [review]
Nice catch @C0urante. LGTM.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java]: Hmm... good point. I think it may make more sense raise a `ConnectException` in order to avoid incorrect/unexpected behavior from the code that's invoking this method. WDYT? Would be interested in @mageshn's thoughts as well

Comment (rayokota) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java]: Yes, it would probably be good to throw an exception here if the method returns false.

**Review by rayokota — APPROVED** [review]
LGTM

**Review by wicknicks — APPROVED** [review]
Thanks, @C0urante. LGTM.

**Comment by C0urante** [comment]
@rhauch @ewencp Hi guys! Got a minor bug fix for you; been reviewed and approved by three contributors and seems ready to be looked at by a committer or two. Would appreciate your review when you get the chance.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java
Comment (kkonstantine) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java]: @C0urante would using a `FutureCallback` be more appropriate here? 
Also do you happen to know what happens if this request is not forwarded to the leader of the group? How much do we care here (not easy to say because this class is missing javadoc)?

You might want to check `completeOrForwardRequest` to see what we do on another call to `herder.connectors` method elsewhere.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java
Comment (kkonstantine) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java]: Have you compared this value with `ConnectorsResource.REQUEST_TIMEOUT_MS`? I know the other one is taking into account rebalances. But how much do we care here? The diff between 90sec and 10sec seems significant.

**Review by kkonstantine — COMMENTED** [review]
Nice find @C0urante !
I think using `FutureCallback` already allows you to do what you implement with the countdown latch. 
Also I have a question on whether we need to forward this request to the leader as we do elsewhere.

**Comment by kkonstantine** [comment]
Forgot to mention that a unit test would be nice (for the whole class even?), since this seems it never worked because it returned immediately.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java]: @kkonstantine thanks for pointing out the `FutureCallback` class, that's much better than the `CountDownLatch`-based approach.

The `ConnectClusterStateImpl` class doesn't have javadocs, but the `ConnectClusterState` interface does. There's no mention of expected behavior if the `ConnectClusterState.connectors()` method is invoked on or off the master node, but FWICT both the standalone and distributed herders rely on their local `ClusterConfigState` snapshots to provide the list of connectors for their own `connectors()` methods, so it doesn't appear forwarding to the leader will be necessary. Let me know if I'm missing anything here; I've tried to go over the code base as best as I can but it's always good to hear from someone more familiar with it.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java]: It looks like the `DistributedHerderRequest` that gets created to fulfill a call to `DistributedHerder.connectors(...)` should complete quickly and won't be blocked even during rebalance. I'm unclear if other requests that are queued up before it could cause a call to `DistributedHerder.connectors(...)` to still take longer than ten seconds but I'm leaning toward no...

**Comment by C0urante** [comment]
@kkonstantine I think I've addressed your two inline comments, and I've added a unit test for the `ConnectClusterStateImpl.connectors()` method. The other method (`connectorHealth(...)`) is synchronous and doesn't really look like it warrants an additional test.

As far as the ten-second timeout goes, I'm open to suggestions but it doesn't appear that rebalances or forwarding to master should be a factor.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImplTest.java
Comment (kkonstantine) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImplTest.java]: nit: my impression is that we don't name the mocks as `mockSomething`. We just declare them/initialize them. Also, I think the annotation style is preferable with easymock/powermock, but I don't feel strong about it.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImplTest.java
Comment (kkonstantine) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImplTest.java:58]: Thanks for simplifying the tests @C0urante! Definitely on the right track. 
As you can see in https://github.com/apache/kafka/blob/trunk/connect/runtime/src/test/java/org/apache/kafka/connect/storage/KafkaConfigBackingStoreTest.java 
you can use the already existing `TestFuture` and be a bit more specific on what to capture and what you expect. 

For example this could be written as: 
```java
    @Test
    public void connectors() {
        Capture<Callback<Collection<String>>> callback = EasyMock.newCapture();
        mockHerder.connectors(EasyMock.capture(callback));
        EasyMock.expectLastCall().andAnswer(new IAnswer<Future<Collection<String>>>() {
            @Override
            public Future<Collection<String>> answer() throws Throwable {
                TestFuture<Collection<String>> future = new TestFuture<>();
                callback.getValue().onCompletion(null, expectedConnectors);
                future.resolveOnGet(expectedConnectors);
                return future;
            }
        });
        EasyMock.replay(mockHerder);
        assertEquals(expectedConnectors, connectClusterState.connectors());
    }
```
this way you also manage to avoid unchecked casts.

**Review by kkonstantine — COMMENTED** [review]
@C0urante this looks even better now IMO and on par with the rest of the code. 
One nit and one more suggested improvement on the tests. Thanks for including those btw. Definitely on the right track.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java
Comment (kkonstantine) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java]: Timeout to read the log and get a snapshot is 3sec. I agree. I think we are good.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImplTest.java
Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImplTest.java:58]: @kkonstantine definitely prefer the `Capture`-based approach that you've outlined, but I don't think the `TestFuture`/`IAnswer` combination in your example is applicable since the future/callback we're dealing with is being passed to `Herder.connectors(...)`, instead of being returned by it.

**Comment by C0urante** [comment]
@kkonstantine thanks for the review! I've taken your comments into account and pushed the next iteration, which consists of changes to how the `ConnectClusterStateImpl` is tested. Hope this looks good and addresses your review remarks, but will be happy to keep iterating if not or if there's anything else that needs work.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImplTest.java
Comment (kkonstantine) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImplTest.java:58]: That's correct. It's called in the method you test. I was carried away in the effort to remove the unchecked casts.

**Review by kkonstantine — APPROVED** [review]
LGTM!

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImpl.java]: Nit: What about `TimeUnit.SECONDS.toMillis(10)`?

**Review by rhauch — COMMENTED** [review]
Nice find, @C0urante! One really minor suggestion for slight improvement in readability, but not a bit deal.

This has never worked, right? This PR should be relatively easy to backport.

**Comment by C0urante** [comment]
@rhauch yep, this bug has been present since REST extensions were added to Connect. I've added your suggested change to the PR, let me know if there are any other alterations you'd like.

**Review by rhauch — APPROVED** [review]
Thanks, @C0urante! LGTM.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImplTest.java
Comment (ewencp) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/health/ConnectClusterStateImplTest.java]: We're moving away from using this since we can use the new `assertThrows` that junit 4.13 supports, which is better since it isolates to the specific call that should fail and allows you to capture and make further assertions. At least for any new tests, we should prefer `assertThrows`, though that does affect ability to backport (since https://github.com/apache/kafka/pull/6248 was only merged to `master`, I think after `2.2` was cut).

**Review by ewencp — COMMENTED** [review]
Mostly looks good, one comment. This is the sort of thing that should come with list of backports -- the bug appears to already be tagged with at least some info about previous versions that are also affected so we probably don't want to close without backporting.

Regarding timeout, it should really be as conservative as possible without being longer than the minimum of (rebalance timeout, default REST API request timeout).

**Comment by C0urante** [comment]
@ewencp can you shed some light on why we'd want to take rebalances into account for the timeout? I could be wrong, but I don't think a rebalance would affect the latency of a call to `Herder.connectors(...)`.

**Comment by ewencp** [comment]
Calling `connectors()` for `DistributedHerder` adds a request to the request queue, which are processed serially in the thread that manages cluster membership so they get a consistent view of the state of the Herder. Pretty much all the requests follow that pattern, though it is true some could potentially be optimized to not require it if some other locking were put in place for certain parts of the state.

The thread that processes the requests is also the one managing the `member` and calling `poll()` on it, which can block on ensuring membership in the group, which means request processing latency can be affected by rebalances.

**Comment by C0urante** [comment]
@ewencp thanks for the clarification; I did not realize that membership and request handling were taken care of with the same thread. I've taken your suggestion literally and just made the timeout the minimum of the timeout for the `/connectors` resource and the rebalance timeout (if running in distributed mode). LMK how this looks!

**Comment by C0urante** [comment]
BTW, I think we want to backport this to 2.0, which according to [the KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-285%3A+Connect+Rest+Extension+Plugin) is when Connect REST extensions were introduced and AFAICT the code for the `ConnectClusterStateImpl` class hasn't been modified since.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java]: nit: i don't love `instanceof` solutions. might be better to generalize thee rebalance timeout so support a small/zero value for `StandaloneConfig`.

I wouldn't block merging on this, but this is an anti-pattern, so if there is a cleaner solution, it might be worth exploring.

**Review by ewencp — APPROVED** [review]
one comment that we could follow up on, but i'm fine merging as is

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java]: @C0urante, what do you think about doing something like the following instead of the `instanceof` block?

```
        Integer rebalanceTimeoutMs = config.getInt(DistributedConfig.REBALANCE_TIMEOUT_MS_CONFIG);
        if (rebalanceTimeoutMs != null) {
            herderRequestTimeoutMs = Math.min(herderRequestTimeoutMs, rebalanceTimeoutMs.longValue());
        }
```
This still uses the `DistributedConfig.REBALANCE_TIMEOUT_MS_CONFIG`, but otherwise is simply checking for the existence of that property to determine if `herderRequestTimeoutMs` should be set to a new value.

Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/RestServer.java]: @rhauch sgtm! I've added your code snippet verbatim to the PR

**Pull request merged**
